extensions/replication/utils/ReplicationTaskScheduler.js

```diff
@@ -1,24 +1,57 @@
 const async = require('async');
 
+const ObjectQueueEntry = require('../../../lib/models/ObjectQueueEntry');
+const ActionQueueEntry = require('../../../lib/models/ActionQueueEntry');
+
 /**
- * Schedule replication tasks for replicated buckets, by ensuring that
- * all updates to the same object are serialized
-*/
+ * Schedule replication tasks by ensuring that all updates to the same object
+ * are serialized
+ */
 class ReplicationTaskScheduler {
     constructor(processingFunc) {
         this._processingFunc = processingFunc;
         this._runningTasksByMasterKey = {};
+        this._runningTasks = {};
+    }
+
+    _getTaskKey(ctx) {
+        const { entry } = ctx;
+        if (entry instanceof ObjectQueueEntry) {
+            const key = entry.getCanonicalKey();
+            const version = entry.getVersionId();
+            const contentMd5 = entry.getContentMd5();
+            return `${key}:${version || ''}:${contentMd5}`;
+        }
+        if (entry instanceof ActionQueueEntry) {
+            const { key, version, contentMd5 } = entry.getAttribute('target');
+            return `${key}:${version || ''}:${contentMd5}`;
+        }
+        return undefined;
+    }
+
+    _getNewMasterKeyQueue(ctx, masterKey) {
+        const queue = async.queue(this._processingFunc);
+        queue.drain = () => delete this._runningTasksByMasterKey[masterKey];
+        this._runningTasksByMasterKey[masterKey] = queue;
+        return queue;
+    }
+
+    _getMasterKeyQueue(ctx, masterKey) {
+        const queue = this._runningTasksByMasterKey[masterKey];
+        return queue || this._getNewMasterKeyQueue(ctx, masterKey);
     }
 
     push(ctx, masterKey, done) {
-        let queue = this._runningTasksByMasterKey[masterKey];
-        if (queue === undefined) {
-            queue = async.queue(this._processingFunc);
-            queue.drain =
-                () => delete this._runningTasksByMasterKey[masterKey];
-            this._runningTasksByMasterKey[masterKey] = queue;
+        const taskKey = this._getTaskKey(ctx);
+        if (this._runningTasks[taskKey]) {
+            return done();
         }
-        queue.push(ctx, done);
+        this._runningTasks[taskKey] = true;
```

Comment:
maybe won't happen because I think entries will be either ObjectQueueEntry or ActionQueueEntry, but `taskKey` can be undefined based on return from `_getTaskKey`

Reply:
Yeah, this shouldn't happen. But I agree it's better to be safe. I'll add a check for that.

Reply:
Added a test for that case as well.